### PR TITLE
feat: forward arbitrary query params to OIDC authorization endpoint

### PR DESCRIPTION
**Pull request by FrankApiyo**
## Summary

`OpenIDClient.login()` and the login viewset action now forward any safe query parameter to the configured OIDC authorization endpoint. Works with any OIDC provider — spec-standard pass-through params and provider-specific hints flow through the same plumbing without per-provider code in ona-oidc.

## Why

Threading individual parameters through ona-oidc as named kwargs (one per parameter) doesn't scale: every provider has its own pass-through hints, and even the OIDC spec defines several. A single `extra_params` dict lets the SPA / Django consumer pass anything the IdP supports without a code change here.

Examples consumers can now use freely:

| Param | Where it comes from |
|---|---|
| `prompt`, `login_hint`, `ui_locales`, `acr_values`, `max_age`, `display`, `id_token_hint` | OIDC Core 1.0 (works everywhere) |
| `kc_idp_hint` | Keycloak — skip the broker IdP-selector page |
| `connection`, `screen_hint`, `organization` | Auth0 |
| `identity_provider` | Amazon Cognito |
| `domain_hint` | Azure AD |
| `hd` | Google |

ona-oidc doesn't need to know about any of them.

## API change

`OpenIDClient.login(redirect_after=None, extra_params=None)`

- `extra_params: Optional[Mapping[str, str]]` — appended to the authorize URL via `urllib.parse.urlencode`.
- New module-level `RESERVED_AUTHORIZE_PARAMS = frozenset({client_id, redirect_uri, scope, response_type, response_mode, state, nonce, code_challenge, code_challenge_method})` silently drops caller-supplied entries that would override the OIDC params ona-oidc itself manages. This is the security guard against a query-string-injection attempt that tries to swap out the redirect URI or the PKCE challenge.
- Falsy `None` / `{}` extras leave the URL unchanged — no stray `&`.

## Viewset behaviour

`BaseOpenIDConnectViewset.login` builds `extra_params` from `request.query_params`, excluding only `next` (which ona-oidc continues to consume for redirect-after-auth caching). Reserved-param filtering lives at the client boundary so it can't be bypassed by future viewset refactors.

Worked example:

```
GET /oidc/<server>/login?prompt=login&login_hint=alice%40example.com&kc_idp_hint=onadata&next=/dashboard
```

→ ona-oidc redirects to:

```
<authorization_endpoint>?client_id=…&redirect_uri=…&scope=…&response_type=…&response_mode=…&prompt=login&login_hint=alice%40example.com&kc_idp_hint=onadata&nonce=…
```

…and the original `next=/dashboard` is cached server-side keyed by the nonce, restored on callback as before.

## Tests

All 76 tests pass (74 pre-existing + 5 new):

- `test_client.test_login_forwards_extra_params` — URL-encoded multi-key forwarding alongside the existing nonce/cache contract.
- `test_client.test_login_no_extra_params_keeps_url_clean` — `None` and `{}` edge cases.
- `test_client.test_login_drops_reserved_authorize_params` — the security guard at the client layer (the right place to test it; viewset-level coverage would just exercise the same code path with extra DRF plumbing).
- `test_viewsets.test_login_forwards_arbitrary_query_params` — multi-key forwarding through the DRF action end-to-end.
- `test_viewsets.test_login_consumes_next_does_not_forward_it` — locks in the viewset's ownership of `next`.

## Test plan

- [x] `python manage.py test tests` passes (76/76) under Django 5.2.11 + Python 3.13.

### CODE REVIEW

**Review by ukanga — CHANGES_REQUESTED** [review]
**Findings**
1. **HIGH** `oidc/viewsets.py:154-161` forwards every browser query param from the public GET login endpoint into the authorize URL. That turns the login URL into a generic auth-parameter tunnel and can leak values like `login_hint` or other sensitive hints into browser history, proxies, and referer logs.
2. **MEDIUM** `oidc/client.py:30-41,347-354` filters several reserved params, but it does not drop the official OIDC request-object parameters `request` and `request_uri`. Those should be blocked at the client boundary so they cannot bypass the app’s own redirect/nonce/state handling.

**Recommendation**
- Keep the client-side reserved-param filter.
- Replace the viewset’s "forward everything except `next`" behavior with a deployment-specific allowlist from settings, ideally per auth server.
- Do not allow `request` or `request_uri` through the public GET login route.

**Example tests that should pass after the fix**
- `OpenIDClient.login()` drops `request` and `request_uri` from `extra_params`.
- `BaseOpenIDConnectViewset.login` forwards only allowlisted query params.
- `BaseOpenIDConnectViewset.login` still consumes `next` locally and does not forward it.
- The resulting authorize URL matches the exact expected value once the blocked params are removed.

**Security Findings**
- **HIGH** Public GET login URL currently accepts arbitrary query params and forwards them to the IdP.
- **MEDIUM** OIDC request-object parameters are not reserved today and should be filtered.

File: oidc/client.py [hunk 1]
```diff
@@ -301,16 +320,39 @@ def _generate_pkce_code_challenge(self, code_verifier: str) -> str:
...
+        so callers can't override the redirect URI, PKCE challenge,
+        nonce, etc.
         """
         url = self.authorization_endpoint + (
             f"?client_id={self.client_id}&redirect_uri={self.redirect_uri}&"
             f"scope={self.scope}&response_type={self.response_type}&"
             f"response_mode={self.response_mode}"
         )
 
+        if extra_params:
```
Comment (ukanga) [inline oidc/client.py:352]: This URL assembly is starting to get brittle. `login()` now builds the base authorize URL with manual string concatenation, then appends filtered extras with `urlencode()`, and later appends PKCE/nonce params again by hand. That mixes encoding concerns and makes the method harder to reason about as more passthrough params get added.

Since this change is already moving `login()` toward dynamic params, I’d strongly recommend building a single ordered `[(key, value), ...]` list and calling `urlencode()` once at the end. That keeps encoding consistent, avoids future escaping bugs, and makes the reserved-param filtering much easier to follow.

Example shape:
```python
params = [
    ("client_id", self.client_id),
    ("redirect_uri", self.redirect_uri),
    ("scope", self.scope),
    ("response_type", self.response_type),
    ("response_mode", self.response_mode),
]
params.extend((k, v) for k, v in safe_params.items())
# append PKCE / nonce here too
url = f"{self.authorization_endpoint}?{urlencode(params)}"
```

File: oidc/viewsets.py [hunk 2]
```diff
@@ -151,7 +151,15 @@ def _resolve_cookie_secure(self) -> bool:
     def login(self, request: HttpRequest, **kwargs: dict) -> HttpResponse:
         client = self._get_client(auth_server=kwargs.get("auth_server"))
         if client:
-            response = client.login(redirect_after=request.query_params.get("next"))
+            extra_params = {
```
Comment (ukanga) [inline oidc/viewsets.py:158]: The login query-param policy is still spread across the viewset and the client. I’d centralize it in a small resolver, with the allowlist stored per auth server in `OPENID_CONNECT_AUTH_SERVERS`, and have `BaseOpenIDConnectViewset.login()` use that single source of truth before calling `OpenIDClient.login()`.

Suggested shape:
```python
OPENID_CONNECT_AUTH_SERVERS = {
    "default": {
        ...
        "LOGIN_QUERY_PARAM_ALLOWLIST": ["prompt", "ui_locales", "acr_values"],
    }
}
```

Then expose a helper like `get_login_query_param_allowlist(auth_server)` in `oidc.utils` and filter browser query params against it there. That keeps the policy explicit, deployment-specific, and confined to one enforcement point instead of being implicit in the view logic.

I’d keep the client-side reserved-parameter filter as a second boundary, but the browser-facing allowlist should live in settings and be resolved centrally.

**Review by ukanga — DISMISSED** [review]

**Review by ukanga — APPROVED** [review]
